File: main/electron-shared/provider-bridge.mjs
import { getSecureStoreSecret } from './codex-services.mjs'

const OFFICIAL_SOURCE_DOMAINS = [
  'openai.com',
  'platform.openai.com',
  'developers.openai.com',
  'anthropic.com',
  'docs.anthropic.com',
  'ai.google.dev',
  'blog.google',
  'developers.googleblog.com',
]

function extractJsonObject(text) {
  const fencedMatch = text.match(/```(?:json)?\s*([\s\S]*?)\s*```/i)
  const candidate = fencedMatch?.[1] ?? text
  const start = candidate.indexOf('{')
  const end = candidate.lastIndexOf('}')

  if (start === -1 || end === -1 || end <= start) {
    throw new Error('JSON 응답을 찾지 못했습니다.')
  }

  return candidate.slice(start, end + 1)
}

function parseGeminiImageDataUrl(imageDataUrl) {
  const match = String(imageDataUrl).match(/^data:(.+?);base64,(.+)$/)

  if (!match) {
    throw new Error('Gemini 이미지 형식을 해석할 수 없습니다.')
  }

  return {
    mimeType: match[1],
    data: match[2],
  }
}

function readRequiredString(payload, key, errorMessage) {
  const value = payload?.[key]

  if (typeof value !== 'string' || !value.trim()) {
    throw new Error(errorMessage)
  }

  return value
}

function isCompoundModel(model) {
  return model === 'groq/compound' || model === 'groq/compound-mini'
}

function isGptOssModel(model) {
  return model === 'openai/gpt-oss-120b' || model === 'openai/gpt-oss-20b'
}

function supportsVision(model) {
  return (
    model === 'meta-llama/llama-4-maverick-17b-128e-instruct' ||
    model === 'meta-llama/llama-4-scout-17b-16e-instruct'
  )
}

function gptOssTools(enabledTools) {
  const tools = []

  if (enabledTools.includes('web_search')) {
    tools.push({ type: 'browser_search' })
  }

  if (enabledTools.includes('code_interpreter')) {
    tools.push({ type: 'code_interpreter' })
  }

  return tools
}

async function invokeGeminiOperation(state, operation, payload) {
  const apiKey = await getSecureStoreSecret('gemini')
  const endpoint = `${state.gemini.endpoint}/${state.gemini.model}:generateContent`

  if (operation === 'summarize') {
    const prompt = readRequiredString(payload, 'prompt', '요약 프롬프트가 없습니다.')
    const response = await fetch(endpoint, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'x-goog-api-key': apiKey,
      },
      body: JSON.stringify({
        contents: [
          {
            role: 'user',
            parts: [{ text: prompt }],
          },
        ],
      }),
      signal: AbortSignal.timeout(8_000),
    })

    if (!response.ok) {
      throw new Error(`Gemini 호출 실패: ${response.status}`)
    }

    const data = await response.json()
    const text = data.candidates?.[0]?.content?.parts?.map((part) => part.text).join('\n')

    if (!text?.trim()) {
      throw new Error('Gemini 응답을 해석할 수 없습니다.')
    }

    return text.trim()
  }

  if (operation === 'extractTextFromImage') {
    const imageDataUrl = readRequiredString(payload, 'imageDataUrl', '이미지 데이터가 없습니다.')
    const { mimeType, data: imageBase64 } = parseGeminiImageDataUrl(imageDataUrl)
    const response = await fetch(endpoint, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'x-goog-api-key': apiKey,
      },
      body: JSON.stringify({
        contents: [
          {
            role: 'user',
            parts: [
              {
                text: [
                  state.uiLocale === 'en'
                    ? 'Read this image like OCR and extract the visible text as faithfully as possible.'
                    : '이 이미지를 OCR처럼 읽고 보이는 글자를 한국어로 그대로 추출해 주세요.',
                  state.uiLocale === 'en'
                    ? 'Return text only, with no explanation or summary.'
                    : '설명, 요약, 해설 없이 텍스트만 반환하세요.',
                  state.uiLocale === 'en'
                    ? 'Preserve the original line breaks as much as possible.'
                    : '줄바꿈은 원문 구조를 최대한 유지하세요.',
                ].join(' '),
              },
              {
                inline_data: {
                  mime_type: mimeType,
                  data: imageBase64,
                },
              },
            ],
          },
        ],
      }),
      signal: AbortSignal.timeout(20_000),
    })

    if (!response.ok) {
      const errorText = await response.text()

      if (response.status === 429) {
        throw new Error('Gemini 할당량을 초과했습니다.')
      }

      throw new Error(`Gemini 이미지 인식 실패: ${response.status} ${errorText}`.trim())
    }

    const data = await response.json()
    const text = data.candidates?.[0]?.content?.parts?.map((part) => part.text).join('\n')

    if (!text?.trim()) {
      throw new Error('Gemini 이미지 인식 응답이 비어 있습니다.')
    }

    return text.trim()
  }

  if (operation === 'verifyFreshness') {
    const text = readRequiredString(payload, 'text', '검증할 텍스트가 없습니다.')
    const response = await fetch(endpoint, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'x-goog-api-key': apiKey,
      },
      body: JSON.stringify({
        contents: [
          {
            role: 'user',
            parts: [
              {
                text: [
                  state.uiLocale === 'en'
                    ? 'Verify only claims about AI models, versions, deprecation status, or whether they are current flagship models.'
                    : '다음 문장에서 AI 모델, 버전, deprecated 여부, 현재 주력 모델 여부와 관련된 주장만 검증하세요.',
                  state.uiLocale === 'en'
                    ? 'You must use web search and prioritize official documentation, official release notes, and official product pages.'
                    : '반드시 웹 검색을 사용하고 공식 문서, 공식 릴리스 노트, 공식 제품 페이지를 우선하세요.',
                  state.uiLocale === 'en'
                    ? 'Use confirmed_outdated or confirmed_current only when the claim is clearly verified. Otherwise use inconclusive.'
                    : '확실히 확인된 경우에만 confirmed_outdated 또는 confirmed_current를 사용하고, 애매하면 inconclusive를 사용하세요.',
                  state.uiLocale === 'en' ? 'Return only the JSON below.' : '반드시 아래 JSON만 반환하세요.',
                  state.uiLocale === 'en'
                    ? '{"status":"confirmed_outdated|confirmed_current|inconclusive","summary":"one short English sentence","checkedClaims":["..."],"references":[{"title":"...","url":"https://..."}]}'
                    : '{"status":"confirmed_outdated|confirmed_current|inconclusive","summary":"짧은 한국어 1문장","checkedClaims":["..."],"references":[{"title":"...","url":"https://..."}]}',
                  '',
                  text.slice(0, 2400),
                ].join('\n'),
              },
            ],
          },
        ],
        tools: [{ google_search: {} }],
      }),
      signal: AbortSignal.timeout(15_000),
    })

    if (!response.ok) {
      const errorText = await response.text()

      if (response.status === 429) {
        throw new Error('Gemini 할당량을 초과했습니다.')
      }

      throw new Error(`Gemini 최신성 검증 실패: ${response.status} ${errorText}`.trim())
    }

    const data = await response.json()
    const content = data.candidates?.[0]?.content?.parts?.map((part) => part.text ?? '').join('\n').trim()

    if (!content) {
      throw new Error('Gemini 최신성 검증 응답이 비어 있습니다.')
    }

    const parsed = JSON.parse(extractJsonObject(content))

    return {
      status: parsed.status ?? 'inconclusive',
      messageKey: 'provider',
      providerSummaryLocale: state.uiLocale,
      providerSummaryText:
        parsed.summary?.trim() ||
        (state.uiLocale === 'en'
          ? 'The web freshness result could not be interpreted.'
          : '웹 검색 기반 최신성 검증 결과를 해석하지 못했습니다.'),
      summary: parsed.summary?.trim() || '웹 검색 기반 최신성 검증 결과를 해석하지 못했습니다.',
      checkedClaims: Array.isArray(parsed.checkedClaims) ? parsed.checkedClaims.slice(0, 5) : [],
      references: Array.isArray(parsed.references) ? parsed.references.slice(0, 3) : [],
    }
  }

  throw new Error('지원하지 않는 Gemini 작업입니다.')
}

async function invokeGroqOperation(state, operation, payload) {
  const apiKey = await getSecureStoreSecret('groq')
  const selectedModel = state.groq.model.trim() || 'groq/compound'

  if (operation === 'summarize') {
    const prompt = readRequiredString(payload, 'prompt', '요약 프롬프트가 없습니다.')
    const headers = {
      Authorization: `Bearer ${apiKey}`,
      'Content-Type': 'application/json',
    }
    const body = {
      model: selectedModel,
      messages: [
        {
          role: 'system',
          content:
            state.uiLocale === 'en'
              ? 'You are K-WarningCheck’s explanation assistant. Keep the judgment intact and answer in exactly one English sentence.'
              : 'K-워닝체크의 보조 요약기입니다. 판정 기준을 바꾸지 말고 한국어 1문장으로만 답하세요.',
        },
        {
          role: 'user',
          content: prompt,
        },
      ],
      temperature: 0.1,
    }

    if (isCompoundModel(selectedModel)) {
      headers['Groq-Model-Version'] = 'latest'
      body.compound_custom = {
        tools: {
          enabled_tools: state.groq.enabledTools,
        },
      }
    } else if (isGptOssModel(selectedModel)) {
      const tools = gptOssTools(state.groq.enabledTools)
      if (tools.length > 0) {
        body.tools = tools
      }
    }

    const response = await fetch(`${state.groq.endpoint}/chat/completions`, {
      method: 'POST',
      headers,
      body: JSON.stringify(body),
      signal: AbortSignal.timeout(8_000),
    })

    const data = await response.json()

    if (!response.ok) {
      throw new Error(data.error?.message ?? `Groq 호출 실패: ${response.status}`)
    }

    const text = data.choices?.[0]?.message?.content

    if (!text?.trim()) {
      throw new Error('Groq 응답을 해석할 수 없습니다.')
    }

    return text.trim()
  }

  if (operation === 'extractTextFromImage') {
    const imageDataUrl = readRequiredString(payload, 'imageDataUrl', '이미지 데이터가 없습니다.')
    const model = supportsVision(selectedModel)
      ? selectedModel
      : 'meta-llama/llama-4-scout-17b-16e-instruct'
    const response = await fetch(`${state.groq.endpoint}/chat/completions`, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${apiKey}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        model,
        messages: [
          {
            role: 'user',
            content: [
              {
                type: 'text',
                text: [
                  state.uiLocale === 'en'
                    ? 'Read this image like OCR and extract the visible text as faithfully as possible.'
                    : '이 이미지를 OCR처럼 읽고 보이는 글자를 한국어로 그대로 추출해 주세요.',
                  state.uiLocale === 'en'
                    ? 'Return text only, with no explanation or summary.'
                    : '설명, 요약, 해설 없이 텍스트만 반환하세요.',
                  state.uiLocale === 'en'
                    ? 'Preserve the original line breaks as much as possible.'
                    : '줄바꿈은 원문 구조를 최대한 유지하세요.',
                ].join(' '),
              },
              {
                type: 'image_url',
                image_url: {
                  url: imageDataUrl,
                },
              },
            ],
          },
        ],
        temperature: 0,
        max_completion_tokens: 2048,
      }),
      signal: AbortSignal.timeout(20_000),
    })

    const data = await response.json()

    if (!response.ok) {
      throw new Error(
        data.error?.message ??
          `Groq 이미지 인식 실패: ${response.status}${model !== selectedModel ? ` (${model} 사용)` : ''}`,
      )
    }

    const text = data.choices?.[0]?.message?.content

    if (!text?.trim()) {
      throw new Error('Groq 이미지 인식 응답이 비어 있습니다.')
    }

    return text.trim()
  }

  if (operation === 'verifyFreshness') {
    const text = readRequiredString(payload, 'text', '검증할 텍스트가 없습니다.')

    if (!(isCompoundModel(selectedModel) || isGptOssModel(selectedModel))) {
      throw new Error('현재 선택한 Groq 모델은 웹 검색 최신성 검증을 지원하지 않습니다.')
    }

    const headers = {
      Authorization: `Bearer ${apiKey}`,
      'Content-Type': 'application/json',
    }
    const body = {
      model: selectedModel,
      messages: [
        {
          role: 'system',
          content: [
            state.uiLocale === 'en'
              ? 'You verify freshness claims about AI models and services.'
              : '당신은 AI 모델·서비스 최신성 검증기입니다.',
            state.uiLocale === 'en'
              ? 'You must use web search and prioritize official documentation, official release notes, and official product pages.'
              : '반드시 웹 검색을 사용해 공식 문서, 공식 릴리스 노트, 공식 제품 페이지를 우선 확인하세요.',
            state.uiLocale === 'en' ? 'Return JSON only.' : '반드시 JSON만 반환하세요.',
          ].join(' '),
        },
        {
          role: 'user',
          content: [
            state.uiLocale === 'en'
              ? 'Verify only claims about AI model names, versions, support status, deprecation, or whether they are currently flagship models.'
              : '다음 문장에서 AI 모델, 버전, 지원 상태, deprecated 여부, 현재 주력 모델 여부와 관련된 주장만 검증하세요.',
            state.uiLocale === 'en'
              ? 'Use confirmed_outdated or confirmed_current only when the claim is clearly verified. Otherwise use inconclusive.'
              : '확실히 확인된 경우에만 confirmed_outdated 또는 confirmed_current를 사용하고, 애매하면 inconclusive를 사용하세요.',
            state.uiLocale === 'en'
              ? 'Return only the JSON schema below.'
              : '반드시 아래 JSON 스키마만 반환하세요.',
            state.uiLocale === 'en'
              ? '{"status":"confirmed_outdated|confirmed_current|inconclusive","summary":"one short English sentence","checkedClaims":["..."],"references":[{"title":"...","url":"https://..."}]}'
              : '{"status":"confirmed_outdated|confirmed_current|inconclusive","summary":"짧은 한국어 1문장","checkedClaims":["..."],"references":[{"title":"...","url":"https://..."}]}',
            '',
            text.slice(0, 2400),
          ].join('\n'),
        },
      ],
      temperature: 0,
    }

    if (isCompoundModel(selectedModel)) {
      headers['Groq-Model-Version'] = 'latest'
      body.compound_custom = {
        tools: {
          enabled_tools: ['web_search'],
        },
      }
      body.search_settings = {
        include_domains: OFFICIAL_SOURCE_DOMAINS,
      }
    } else {
      body.tools = [{ type: 'browser_search' }]
      body.tool_choice = 'required'
      body.search_settings = {
        include_domains: OFFICIAL_SOURCE_DOMAINS,
      }
    }

    const response = await fetch(`${state.groq.endpoint}/chat/completions`, {
      method: 'POST',
      headers,
      body: JSON.stringify(body),
      signal: AbortSignal.timeout(15_000),
    })

    const data = await response.json()

    if (!response.ok) {
      throw new Error(data.error?.message ?? `Groq 최신성 검증 실패: ${response.status}`)
    }

    const content = data.choices?.[0]?.message?.content?.trim()

    if (!content) {
      throw new Error('Groq 최신성 검증 응답이 비어 있습니다.')
    }

    const parsed = JSON.parse(extractJsonObject(content))

    return {
      status: parsed.status ?? 'inconclusive',
      messageKey: 'provider',
      providerSummaryLocale: state.uiLocale,
      providerSummaryText:
        parsed.summary?.trim() ||
        (state.uiLocale === 'en'
          ? 'The web freshness result could not be interpreted.'
          : '웹 검색 기반 최신성 검증 결과를 해석하지 못했습니다.'),
      summary: parsed.summary?.trim() || '웹 검색 기반 최신성 검증 결과를 해석하지 못했습니다.',
      checkedClaims: Array.isArray(parsed.checkedClaims) ? parsed.checkedClaims.slice(0, 5) : [],
      references: Array.isArray(parsed.references) ? parsed.references.slice(0, 3) : [],
    }
  }

  throw new Error('지원하지 않는 Groq 작업입니다.')
}

export async function invokeProviderBridge({ store, provider, operation, payload }) {
  const state = await store.getProviderState()

  if (provider === 'gemini') {
    return invokeGeminiOperation(state, operation, payload ?? {})
  }

  if (provider === 'groq') {
    return invokeGroqOperation(state, operation, payload ?? {})
  }

  throw new Error('지원하지 않는 provider bridge 요청입니다.')
}

File: tauri-app/src/commands/provider_bridge.rs
use reqwest::Client;
use serde_json::{json, Value};
use std::time::Duration;

use crate::commands::provider_state::get_provider_state_merged;
use crate::secure::get_secure_store_secret;

const OFFICIAL_SOURCE_DOMAINS: &[&str] = &[
    "openai.com",
    "platform.openai.com",
    "developers.openai.com",
    "anthropic.com",
    "docs.anthropic.com",
    "ai.google.dev",
    "blog.google",
    "developers.googleblog.com",
];

fn read_required_string(payload: &Value, key: &str, error: &str) -> Result<String, String> {
    payload
        .get(key)
        .and_then(|v| v.as_str())
        .filter(|s| !s.trim().is_empty())
        .map(|s| s.to_string())
        .ok_or_else(|| error.to_string())
}

fn read_bool(payload: &Value, key: &str) -> bool {
    payload
        .get(key)
        .and_then(|v| v.as_bool())
        .unwrap_or(false)
}

fn is_compound_model(model: &str) -> bool {
    model == "groq/compound" || model == "groq/compound-mini"
}

fn is_gpt_oss_model(model: &str) -> bool {
    model == "openai/gpt-oss-120b" || model == "openai/gpt-oss-20b"
}

fn supports_vision(model: &str) -> bool {
    model == "meta-llama/llama-4-maverick-17b-128e-instruct"
        || model == "meta-llama/llama-4-scout-17b-16e-instruct"
}

fn gemini_endpoint(state: &Value) -> String {
    let endpoint = state
        .get("gemini")
        .and_then(|g| g.get("endpoint"))
        .and_then(|v| v.as_str())
        .unwrap_or("https://generativelanguage.googleapis.com/v1beta/models");
    let model = state
        .get("gemini")
        .and_then(|g| g.get("model"))
        .and_then(|v| v.as_str())
        .unwrap_or("gemini-3.1-flash-lite-preview");
    format!("{endpoint}/{model}:generateContent")
}

fn groq_endpoint(state: &Value) -> String {
    let endpoint = state
        .get("groq")
        .and_then(|g| g.get("endpoint"))
        .and_then(|v| v.as_str())
        .unwrap_or("https://api.groq.com/openai/v1");
    format!("{endpoint}/chat/completions")
}

fn groq_model(state: &Value) -> String {
    state
        .get("groq")
        .and_then(|g| g.get("model"))
        .and_then(|v| v.as_str())
        .unwrap_or("groq/compound-mini")
        .trim()
        .to_string()
}

fn gemini_thinking_config(model: &str) -> Value {
    if model.starts_with("gemini-3") {
        if model.contains("flash-lite") || model.contains("flash") {
            json!({ "thinkingLevel": "minimal" })
        } else {
            json!({ "thinkingLevel": "low" })
        }
    } else if model.starts_with("gemini-2.5") {
        json!({ "thinkingBudget": 0 })
    } else {
        Value::Null
    }
}

fn parse_data_url(url: &str) -> Result<(String, String), String> {
    let prefix = "data:";
    if !url.starts_with(prefix) {
        return Err("이미지 형식을 해석할 수 없습니다.".into());
    }
    let rest = &url[prefix.len()..];
    let semi = rest.find(';').ok_or("이미지 형식을 해석할 수 없습니다.")?;
    let mime = &rest[..semi];
    let after_semi = &rest[semi + 1..];
    if !after_semi.starts_with("base64,") {
        return Err("이미지 형식을 해석할 수 없습니다.".into());
    }
    let data = &after_semi[7..];
    Ok((mime.to_string(), data.to_string()))
}

fn extract_gemini_text(data: &Value) -> String {
    data.pointer("/candidates/0/content/parts")
        .and_then(|p| p.as_array())
        .map(|parts| {
            parts
                .iter()
                .filter_map(|p| p.get("text").and_then(|t| t.as_str()))
                .collect::<Vec<_>>()
                .join("\n")
        })
        .unwrap_or_default()
        .trim()
        .to_string()
}

fn extract_groq_text(data: &Value) -> String {
    data.pointer("/choices/0/message/content")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .trim()
        .to_string()
}

async fn invoke_gemini(state: &Value, operation: &str, payload: &Value) -> Result<Value, String> {
    if operation != "analyzeRisk" {
        return Err("지원하지 않는 Gemini 작업입니다.".into());
    }

    let api_key = get_secure_store_secret("gemini").await?;
    let endpoint = gemini_endpoint(state);
    let model = state
        .get("gemini")
        .and_then(|g| g.get("model"))
        .and_then(|v| v.as_str())
        .unwrap_or("gemini-3.1-flash-lite-preview")
        .trim()
        .to_string();
    let client = Client::new();
    let prompt = read_required_string(payload, "prompt", "분석 프롬프트가 없습니다.")?;
    let image_data_url = payload
        .get("imageDataUrl")
        .and_then(|v| v.as_str())
        .map(|value| value.trim())
        .filter(|value| !value.is_empty());
    let mut parts = vec![json!({ "text": prompt })];

    if let Some(image_data_url) = image_data_url {
        let (mime_type, data) = parse_data_url(image_data_url)?;
        parts.push(json!({
            "inline_data": {
                "mime_type": mime_type,
                "data": data,
            }
        }));
    }

    let use_web_search = read_bool(payload, "useWebSearch");
    let non_search_max_output = if model.contains("pro") { 420 } else { 320 };
    let mut schema_properties = json!({
        "summary": {
            "type": "string",
            "description": "짧은 1문장"
        },
        "responseText": {
            "type": "string",
            "description": "최대 2개의 짧은 문장"
        },
        "evidence": {
            "type": "array",
            "minItems": 1,
            "maxItems": 2,
            "items": { "type": "string" }
        }
    });
    if use_web_search {
        schema_properties["freshnessNote"] = json!({ "type": ["string", "null"] });
    }
    if image_data_url.is_some() {
        schema_properties["extractedText"] = json!({ "type": ["string", "null"] });
    }
    let thinking_config = gemini_thinking_config(&model);
    let resp = client
        .post(&endpoint)
        .header("Content-Type", "application/json")
        .header("x-goog-api-key", &api_key)
        .timeout(Duration::from_secs(if use_web_search { 35 } else { 15 }))
        .json(&json!({
            "contents": [{
                "role": "user",
                "parts": parts
            }],
            "tools": if use_web_search { json!([{ "google_search": {} }]) } else { json!([]) },
            "generationConfig": {
                "temperature": 0.1,
                "maxOutputTokens": if use_web_search || image_data_url.is_some() { 420 } else { non_search_max_output },
                "responseMimeType": "application/json",
                "responseJsonSchema": {
                    "type": "object",
                    "properties": schema_properties,
                    "required": ["summary", "responseText", "evidence"]
                },
                "thinkingConfig": thinking_config
            }
        }))
        .send()
        .await
        .map_err(|e| format!("Gemini 분석 실패: {e}"))?;

    let status = resp.status();
    if status.as_u16() == 429 {
        return Err("Gemini 할당량을 초과했습니다.".into());
    }
    if !status.is_success() {
        let error_text = resp.text().await.unwrap_or_default();
        return Err(format!("Gemini 분석 실패: {status} {error_text}"));
    }

    let data: Value = resp.json().await.map_err(|e| e.to_string())?;
    let text = extract_gemini_text(&data);
    if text.is_empty() {
        return Err("Gemini 분석 응답이 비어 있습니다.".into());
    }

    Ok(json!(text))
}

async fn invoke_groq(state: &Value, operation: &str, payload: &Value) -> Result<Value, String> {
    if operation != "analyzeRisk" {
        return Err("지원하지 않는 Groq 작업입니다.".into());
    }

    let api_key = get_secure_store_secret("groq").await?;
    let selected_model = groq_model(state);
    let prompt = read_required_string(payload, "prompt", "분석 프롬프트가 없습니다.")?;
    let image_data_url = payload
        .get("imageDataUrl")
        .and_then(|v| v.as_str())
        .map(|value| value.trim())
        .filter(|value| !value.is_empty());
    let use_vision = image_data_url.is_some();
    let model = if use_vision {
        if supports_vision(&selected_model) {
            selected_model.clone()
        } else {
            "meta-llama/llama-4-scout-17b-16e-instruct".to_string()
        }
    } else if selected_model.is_empty() {
        "groq/compound-mini".to_string()
    } else {
        selected_model.clone()
    };
    let use_web_search = read_bool(payload, "useWebSearch") && !use_vision;
    let client = Client::new();
    let mut headers = reqwest::header::HeaderMap::new();
    headers.insert("Authorization", format!("Bearer {api_key}").parse().unwrap());
    headers.insert("Content-Type", "application/json".parse().unwrap());

    let message_content = if let Some(image_data_url) = image_data_url {
        json!([
            { "type": "text", "text": prompt },
            { "type": "image_url", "image_url": { "url": image_data_url } }
        ])
    } else {
        json!(prompt)
    };

    let mut body = json!({
        "model": model,
        "messages": [
            {
                "role": "user",
                "content": message_content
            }
        ],
        "temperature": 0.1,
        "max_completion_tokens": if use_web_search || use_vision { 700 } else { 240 },
        "response_format": { "type": "json_object" }
    });

    if is_compound_model(&model) {
        body["compound_custom"] = json!({
            "tools": {
                "enabled_tools": if use_web_search { json!(["web_search"]) } else { json!([]) }
            }
        });
    }

    if use_web_search {
        if is_compound_model(&model) {
            headers.insert("Groq-Model-Version", "latest".parse().unwrap());
            body["search_settings"] = json!({
                "include_domains": OFFICIAL_SOURCE_DOMAINS
            });
        } else if is_gpt_oss_model(&model) {
            body["tools"] = json!([{ "type": "browser_search" }]);
            body["tool_choice"] = json!("required");
            body["search_settings"] = json!({
                "include_domains": OFFICIAL_SOURCE_DOMAINS
            });
        }
    }

    let resp = client
        .post(&groq_endpoint(state))
        .headers(headers)
        .timeout(Duration::from_secs(if use_web_search { 30 } else { 20 }))
        .json(&body)
        .send()
        .await
        .map_err(|e| format!("Groq 분석 실패: {e}"))?;

    let data: Value = resp.json().await.map_err(|e| e.to_string())?;
    if let Some(err) = data
        .get("error")
        .and_then(|e| e.get("message"))
        .and_then(|m| m.as_str())
    {
        return Err(err.to_string());
    }

    let text = extract_groq_text(&data);
    if text.is_empty() {
        return Err("Groq 분석 응답이 비어 있습니다.".into());
    }

    Ok(json!(text))
}

#[tauri::command]
pub async fn kwc_provider_bridge_invoke(
    provider: String,
    operation: String,
    payload: Value,
) -> Result<Value, String> {
    let state = get_provider_state_merged().await;
    let payload = if payload.is_null() {
        json!({})
    } else {
        payload
    };

    match provider.as_str() {
        "gemini" => invoke_gemini(&state, &operation, &payload).await,
        "groq" => invoke_groq(&state, &operation, &payload).await,
        _ => Err("지원하지 않는 provider bridge 요청입니다.".into()),
    }
}
